### ISSUE TITLE
Meaning of (mini)batch size has been changed

### ISSUE THREAD

**Issue by jaekyeom** [rlworkgroup/garage#621]
I've noticed that in garage, policies take trajectories as inputs even when they are not recurrent policies. They flatten the samples on the TensorFlow's side.

However, I find that problematic, since it changes the meaning of `batch_size` for the policy optimizer.
For instance, if we look at [examples/tf/ppo_pendulum.py](https://github.com/rlworkgroup/garage/blob/master/examples/tf/ppo_pendulum.py), it passes `optimizer_args=dict(batch_size=32, max_epochs=10)` to `PPO.__init__()`. But what it really means is, "transitions from 32 trajectories will form a single minibatch", not "32 transitions will form a single minibatch".

Correct me if I'm wrong, but my simple `tf.Print` test suggests the same thing (I printed out the shape of `i.flat.action_var` and `i.valid.action_var`).

**Comment by ryanjulian** [comment]
@jaekyeom this is true only for the policy gradient algorithms (those based on `garage/tf/npo.py`). In particular, this would mostly affect PPO, since most other policy gradient algorithms (e.g. TRPO) don't use SGD.

Most people never change the minibatch size in PPO, so we never considered it a big issue.

I'm not sure it's actually "problematic," although it's certainly inconsistent with the definition of `batch_size` in `LocalTfRunner` and throughout the codebase (in which `batch_size` always refers to the number of time steps processed regardless of which trajectory they originated from).

Feeding the trajectories in dense form for policy gradient algorithms allows us to process samples in the GPU much more readily (e.g. this matters if your observations are images or you have a lot of samples). It also unifies the implementation between recurrent and non-recurrent policies. It allows us to provide a gradient path through the entire policy gradient loss including differentiable reward augmentations, for instance for maximum-entropy or entropy-regularized policies.

Changing the batch parameter to be whole trajectories rather than samples should make minibatching less-effective for small batch sizes, and should have a negligible effect for large batches. It does not seem to have a negative effect on performance (our PPO regularly outperforms openai/baselines on Mujoco1M, see https://rlworkgroup.github.io/benchmarks/).

If you feel strongly about the issue, I'd be happy to review a pull request which passes timestep-oriented data to the optimizer in NPO while preserving the dense computation in the loss function. This would require reshaping the input data to be trajectory-oriented in the beginning of the loss function.
